ext/semian/resource.c

```diff
@@ -63,9 +69,20 @@ semian_resource_acquire(int argc, VALUE *argv, VALUE self)
 VALUE
 semian_resource_destroy(VALUE self)
 {
+  struct timespec ts = { 0 };
+  ts.tv_sec = INTERNAL_TIMEOUT;
+
   semian_resource_t *res = NULL;
 
   TypedData_Get_Struct(self, semian_resource_t, &semian_resource_type, res);
+
+  // Prevent a race to deletion
+  if (perform_semop(res->sem_id, SI_SEM_LOCK, -1, 0, &ts) == -1) {
```

Comment:
This is equivalent to sem_meta_lock, but we want to allow EINVAL as that will show us that the resource has already been deleted.